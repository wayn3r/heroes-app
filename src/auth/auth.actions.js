import { types } from '../types/types'

/**
 *
 * @param {String} name
 */
export const login = name => ({
    type: types.login,
    payload: { name },
})
export const logout = () => ({
    type: types.logout,
})
