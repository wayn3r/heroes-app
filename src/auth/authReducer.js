import { types } from '../types/types'

export const initialState = {
    name: '',
    logged: false,
}

export const authReducer = (state = initialState, action = {}) => {
    const { type, payload } = action
    switch (type) {
        case types.login:
            const { name } = payload
            return { name, logged: true }
        case types.logout:
            return { ...initialState }
        default:
            return state
    }
}
