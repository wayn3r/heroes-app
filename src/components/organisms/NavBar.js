import { useContext } from 'react'
import { Link, NavLink, useNavigate } from 'react-router-dom'
import { logout } from '../../auth/auth.actions'
import { AuthContext } from '../../auth/AuthContext'

const NavBar = () => {
    const { user, dispatch } = useContext(AuthContext)
    const navigate = useNavigate()
    const handleLogout = () => {
        dispatch(logout())
        navigate('/login', { replace: true })
    }
    return (
        <nav className="navbar navbar-expand-sm navbar-dark bg-dark px-5">
            <Link className="navbar-brand" to="/">
                Asociaciones
            </Link>

            <div className="navbar-collapse">
                <div className="navbar-nav">
                    <NavLink className="nav-item nav-link" to="/marvel">
                        Marvel
                    </NavLink>

                    <NavLink className="nav-item nav-link" to="/dc">
                        DC
                    </NavLink>

                    <NavLink className="nav-item nav-link" to="/search">
                        Search
                    </NavLink>
                </div>
            </div>

            <div className="navbar-collapse collapse w-100 order-3 dual-collapse2">
                <ul className="navbar-nav ms-auto">
                    <span className="nav-item nav-link text-info">{user.name}</span>
                    <button className="nav-item nav-link btn" onClick={handleLogout}>
                        Logout
                    </button>
                </ul>
            </div>
        </nav>
    )
}

export default NavBar
