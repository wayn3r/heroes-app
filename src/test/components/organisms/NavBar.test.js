import { mount } from 'enzyme'
import { MemoryRouter } from 'react-router-dom'
import { logout } from '../../../auth/auth.actions'
import { AuthContext } from '../../../auth/AuthContext'
import NavBar from '../../../components/organisms/NavBar'
import { loggedUser } from '../../fixins/AuthContextValue'
import { act } from '@testing-library/react'
const mockNavigate = jest.fn()
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'), // use actual for all non-hook parts
    useNavigate: () => mockNavigate,
}))
describe('Pruebas en el <NavBar />', () => {
    const dispatch = jest.fn()

    const wrapper = mount(
        <AuthContext.Provider value={{ user: loggedUser, dispatch }}>
            <MemoryRouter>
                <NavBar />
            </MemoryRouter>
        </AuthContext.Provider>,
    )
    test('debe mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot()
        expect(wrapper.find('.text-info').text().trim()).toBe(loggedUser.name)
    })

    test('debe llamar el dispatch con logout() y redireccionar al login ', () => {
        act(() => {
            wrapper.find('button').prop('onClick')()
        })

        expect(dispatch).toHaveBeenCalledWith(logout())
        expect(mockNavigate).toHaveBeenCalledWith('/login', { replace: true })
    })
})
