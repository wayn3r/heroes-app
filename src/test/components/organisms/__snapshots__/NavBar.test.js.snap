// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el <NavBar /> debe mostrarse correctamente 1`] = `
<nav
  className="navbar navbar-expand-sm navbar-dark bg-dark px-5"
>
  <a
    className="navbar-brand"
    href="/"
    onClick={[Function]}
  >
    Asociaciones
  </a>
  <div
    className="navbar-collapse"
  >
    <div
      className="navbar-nav"
    >
      <a
        className="nav-item nav-link"
        href="/marvel"
        onClick={[Function]}
      >
        Marvel
      </a>
      <a
        className="nav-item nav-link"
        href="/dc"
        onClick={[Function]}
      >
        DC
      </a>
      <a
        className="nav-item nav-link"
        href="/search"
        onClick={[Function]}
      >
        Search
      </a>
    </div>
  </div>
  <div
    className="navbar-collapse collapse w-100 order-3 dual-collapse2"
  >
    <ul
      className="navbar-nav ms-auto"
    >
      <span
        className="nav-item nav-link text-info"
      >
        Wayner
      </span>
      <button
        className="nav-item nav-link btn"
        onClick={[Function]}
      >
        Logout
      </button>
    </ul>
  </div>
</nav>
`;
